File: benchmark/comparative/benchmarks-QuantumToolbox.jl/benchmarkutils.jl
module benchmarkutils

using QuantumToolbox
using JSON

set_zero_subnormals(true)

const rootpath = abspath(".")
const commitID = try
    string(pkgversion(QuantumToolbox))
catch
    "unknown"
end
println("Detected QuantumToolbox.jl version: ", commitID)

benchmark_directory = "benchmarks-QuantumToolbox.jl"
checkvalues_directory = "checks"

function examplename(name)
    if endswith(name, "]")
        return rsplit(name, "[", limit=2)[1]
    end
    name
end

function check(name, D, eps=1e-5)
    check_path = "../checks/$(examplename(name)).json"
    if ispath(check_path)
        println("Checking against check file.")
        data = JSON.parsefile(check_path)
        for (N, result) in D
            r = data[string(N)]
            if isnan(result) || abs(result-r)/abs(r) > eps
                println("Warning: Result may be incorrect in ", name, ": ", result, " <-> ", r)
            end
        end
    else
        println("No check file found.")
    end
end

function save(name, results)
    result_path = "../results/results-QuantumToolbox.jl-$commitID-$name.json"
    f = open(result_path, "w")
    write(f, JSON.json(results))
    close(f)
end

end # module

File: benchmark/comparative/benchmarks-QuTiP/benchmarkutils.py
import timeit
import json
import qutip
import numpy as np

commitID = qutip.__version__
result_path = "../results/results-QuTiP-{}-{}.json"


def examplename(name):
    if name.endswith("]"):
        return name.rsplit("[", 1)[0]
    else:
        return name


def run_benchmark(f, *args, samples=5, evals=1):
    D = {"f": f, "args": args}
    t = timeit.repeat("f(*args)", globals=D, number=evals, repeat=samples)
    return min(t) / evals


def check(name, D, eps=1e-5):
    check_path = "../checks/" + examplename(name) + ".json"
    try:
        with open(check_path) as f:
            data = json.load(f)
        for N, result in D.items():
            r = data[str(N)]
            if np.isnan(result) or abs(result - r) / abs(r) > eps:
                print(f"Warning: Result may be incorrect in {name}: {result} <-> {r}")
    except FileNotFoundError:
        print("No check file found - write results to check file.")
        with open(check_path, "w") as f:
            json.dump(D, f)


def save(name, results):
    path = result_path.format(commitID, name)
    with open(path, "w") as f:
        json.dump(results, f)
